packet.d.ts

```diff
@@ -0,0 +1,193 @@
+  cmd: "connect";
+  duration: number;
+  clientId: string;
+  will: boolean;
+  cleanSession: boolean;
+}
+
+interface ConnAck extends GenericAck {
+  cmd: "connack";
+}
+
+interface WillTopic extends BaseMQTTSNPacket {
+  cmd: "willtopic";
+  qos?: QOS;
+  retain?: boolean;
+  willTopic?: string;
+}
+
+interface WillTopicResp extends GenericAck {
+  cmd: "willtopicresp";
+}
+
+interface WillTopicUpd extends BaseMQTTSNPacket {
+  cmd: "willtopicupd";
+  qos?: QOS;
+  retain?: boolean;
+  willTopic?: string;
+}
+
+interface WillMsg extends BaseMQTTSNPacket {
+  cmd: "willmsg";
+  willMsg: string;
+}
+
+interface WillMsgResp extends GenericAck {
+  cmd: "willmsgresp";
+}
+
+interface WillMsgUpd extends WillMsg {
+  cmd: "willmsgupd";
+}
+
+interface Register extends BaseMQTTSNPacket {
+  cmd: "register";
+  topicId: number;
+  msgId: number;
+  topicName: string;
+}
+
+interface RegAck extends GenericAck {
+  cmd: "regack";
+  topicId: number;
+  msgId: number;
+}
+
+interface Publish<TT extends TopicType, TID> extends BaseMQTTSNPacket {
+  cmd: "publish";
+  dup: boolean;
+  qos: QOS;
+  topicIdType: TT;
+  topicId: TID;
+  msgId: number;
+  payload: Buffer;
+}
+
+type PublishShortTopic = Publish<TopicType.SHORT, string>;
+type PublishPredefined = Publish<TopicType.PREDEFINED, number>;
+type PublishRegisteredTopic = Publish<TopicType.NORMAL, number>;
+
+interface PubAck extends GenericAck {
+  cmd: "puback";
+  topicId: number;
+  msgId: number;
+}
+
+interface PubComp extends BaseMQTTSNPacket {
+  cmd: "pubcomp";
+  msgId: number;
+}
+
+interface PubRec extends BaseMQTTSNPacket {
+  cmd: "pubrec";
+  msgId: number;
+}
+
+interface PubRel extends BaseMQTTSNPacket {
+  cmd: "pubrel";
+  msgId: number;
+}
+
+interface UnsubAck extends BaseMQTTSNPacket {
+  cmd: "unsuback";
+  msgId: number;
+}
+
+interface Unsubscribe<TT extends TopicType, TN, TID> extends BaseMQTTSNPacket {
+  cmd: "unsubscribe";
+  msgId: number;
+  topicIdType: TT;
+  topicName: TN;
+  topicId: TID;
+}
+
+type UnsubscribePredefined = Unsubscribe<TopicType.PREDEFINED, undefined, number>;
+type UnsubscribeRegisteredTopic = Unsubscribe<TopicType.NORMAL, undefined, number>;
+type UnsubscribeShortTopic = Unsubscribe<TopicType.SHORT, string, undefined>;
+
+interface Subscribe<TT extends TopicType, TN, TID> extends Unsubscribe<TT, TN, TID> {
+  cmd: "subscribe";
+  dup: boolean;
+  qos: QOS;
+}
+
+declare type SubscribePredefined = Subscribe<TopicType.PREDEFINED, undefined, number>;
+declare type SubscribeRegisteredTopic = Subscribe<TopicType.NORMAL, undefined, number>;
+declare type SubscribeShortTopic = Subscribe<TopicType.SHORT, string, undefined>;
+
+interface SubAck extends GenericAck {
+  cmd: "suback";
+  topicId: number;
+  msgId: number;
+  qos: QOS;
+}
+
+interface PingReq extends BaseMQTTSNPacket {
+  cmd: "pingreq";
+  clientId?: string;
+}
+
+interface Disconnect extends BaseMQTTSNPacket {
+  cmd: "disconnect";
+  duration?: number;
+}
+
+
+export type MQTTPacket =
```

Comment:
related to previous comment:
```suggestion
export type MQTTSNPacket =
```